File: examples/sdk/python/miniodb_client.py
#!/usr/bin/env python3
"""
MinIODB Python SDK 示例

本示例展示如何通过 Redis Streams 或 Kafka 向 MinIODB 发布数据
适用于第三方系统集成，实现业务实时数据同步

依赖安装:
    pip install redis kafka-python grpcio grpcio-tools

使用方法:
    python miniodb_client.py --mode redis    # 使用 Redis Streams
    python miniodb_client.py --mode kafka    # 使用 Kafka
    python miniodb_client.py --mode grpc     # 使用 gRPC 直连
"""

import json
import time
import uuid
import argparse
from datetime import datetime
from typing import Dict, List, Any, Optional
from dataclasses import dataclass, field, asdict


# ===============================================
# 数据流格式定义（与 MinIODB 保持一致）
# ===============================================

@dataclass
class DataRecord:
    """数据记录"""
    id: str
    timestamp: int  # 毫秒时间戳
    payload: Dict[str, Any]
    version: Optional[int] = None
    deleted_at: Optional[int] = None


@dataclass
class EventMetadata:
    """事件元数据"""
    source: Optional[str] = None
    batch_id: Optional[str] = None
    partition_key: Optional[str] = None
    priority: str = "normal"
    trace_id: Optional[str] = None
    tags: Dict[str, str] = field(default_factory=dict)
    retry_count: int = 0


@dataclass
class DataEvent:
    """数据事件"""
    event_id: str
    event_type: str  # insert, update, delete, batch
    table: str
    timestamp: str  # ISO 8601 格式
    records: List[DataRecord]
    metadata: EventMetadata

    def to_json(self) -> str:
        """序列化为 JSON"""
        data = {
            "event_id": self.event_id,
            "event_type": self.event_type,
            "table": self.table,
            "timestamp": self.timestamp,
            "records": [asdict(r) for r in self.records],
            "metadata": asdict(self.metadata)
        }
        return json.dumps(data)

    @classmethod
    def from_json(cls, json_str: str) -> "DataEvent":
        """从 JSON 反序列化"""
        data = json.loads(json_str)
        records = [DataRecord(**r) for r in data["records"]]
        metadata = EventMetadata(**data["metadata"])
        return cls(
            event_id=data["event_id"],
            event_type=data["event_type"],
            table=data["table"],
            timestamp=data["timestamp"],
            records=records,
            metadata=metadata
        )


def new_event(event_type: str, table: str, records: List[DataRecord], 
              source: str = None) -> DataEvent:
    """创建新的数据事件"""
    event_id = f"evt_{str(uuid.uuid4())[:12]}"
    timestamp = datetime.utcnow().isoformat() + "Z"
    metadata = EventMetadata(source=source)
    return DataEvent(
        event_id=event_id,
        event_type=event_type,
        table=table,
        timestamp=timestamp,
        records=records,
        metadata=metadata
    )


def new_insert_event(table: str, record: DataRecord, source: str = None) -> DataEvent:
    """创建插入事件"""
    return new_event("insert", table, [record], source)


def new_batch_event(table: str, records: List[DataRecord], source: str = None) -> DataEvent:
    """创建批量插入事件"""
    return new_event("batch", table, records, source)


# ===============================================
# Redis Streams 客户端
# ===============================================

class RedisStreamClient:
    """Redis Streams 客户端"""

    def __init__(self, host: str = "localhost", port: int = 6379, 
                 password: str = None, db: int = 0,
                 stream_prefix: str = "miniodb:stream:"):
        import redis
        self.client = redis.Redis(
            host=host, 
            port=port, 
            password=password, 
            db=db,
            decode_responses=True
        )
        self.stream_prefix = stream_prefix
        # 测试连接
        self.client.ping()

    def publish(self, event: DataEvent) -> str:
        """发布事件到 Redis Stream"""
        stream_key = f"{self.stream_prefix}{event.table}"
        fields = {
            "event_id": event.event_id,
            "event_type": event.event_type,
            "table": event.table,
            "timestamp": event.timestamp,
            "data": event.to_json()
        }
        return self.client.xadd(stream_key, fields)

    def publish_batch(self, events: List[DataEvent]) -> List[str]:
        """批量发布事件"""
        pipe = self.client.pipeline()
        for event in events:
            stream_key = f"{self.stream_prefix}{event.table}"
            fields = {
                "event_id": event.event_id,
                "event_type": event.event_type,
                "table": event.table,
                "timestamp": event.timestamp,
                "data": event.to_json()
            }
            pipe.xadd(stream_key, fields)
        return pipe.execute()

    def close(self):
        """关闭连接"""
        self.client.close()


# ===============================================
# Kafka 客户端
# ===============================================

class KafkaClient:
    """Kafka 客户端"""

    def __init__(self, brokers: List[str], topic_prefix: str = "miniodb-"):
        from kafka import KafkaProducer
        self.producer = KafkaProducer(
            bootstrap_servers=brokers,
            value_serializer=lambda v: v.encode('utf-8'),
            key_serializer=lambda k: k.encode('utf-8') if k else None
        )
        self.topic_prefix = topic_prefix

    def publish(self, event: DataEvent):
        """发布事件到 Kafka"""
        topic = f"{self.topic_prefix}{event.table}"
        future = self.producer.send(
            topic,
            key=event.table,
            value=event.to_json(),
            headers=[
                ("event_id", event.event_id.encode()),
                ("event_type", event.event_type.encode())
            ]
        )
        return future.get(timeout=10)

    def publish_batch(self, events: List[DataEvent]):
        """批量发布事件"""
        futures = []
        for event in events:
            topic = f"{self.topic_prefix}{event.table}"
            future = self.producer.send(
                topic,
                key=event.table,
                value=event.to_json(),
                headers=[
                    ("event_id", event.event_id.encode()),
                    ("event_type", event.event_type.encode())
                ]
            )
            futures.append(future)
        
        # 等待所有消息发送完成
        results = []
        for future in futures:
            results.append(future.get(timeout=10))
        return results

    def close(self):
        """关闭连接"""
        self.producer.flush()
        self.producer.close()


# ===============================================
# gRPC 客户端（简化版，需要生成 protobuf 代码）
# ===============================================

class GRPCClient:
    """gRPC 客户端（简化版）
    
    注意: 完整使用需要先生成 protobuf Python 代码:
    python -m grpc_tools.protoc -I../../api/proto --python_out=. --grpc_python_out=. miniodb/v1/miniodb.proto
    """

    def __init__(self, addr: str = "localhost:8080"):
        import grpc
        self.channel = grpc.insecure_channel(addr)
        # 注意: 需要导入生成的 protobuf 代码
        # from miniodb.v1 import miniodb_pb2, miniodb_pb2_grpc
        # self.stub = miniodb_pb2_grpc.MinIODBServiceStub(self.channel)
        print("注意: gRPC 模式需要先生成 protobuf Python 代码")

    def write_data(self, table: str, record: DataRecord):
        """写入数据"""
        # 需要 protobuf 代码
        raise NotImplementedError("需要先生成 protobuf Python 代码")

    def close(self):
        """关闭连接"""
        self.channel.close()


# ===============================================
# 主程序
# ===============================================

def generate_record(index: int) -> DataRecord:
    """生成测试数据记录"""
    return DataRecord(
        id=f"record_{int(time.time() * 1000)}_{index}",
        timestamp=int(time.time() * 1000),
        payload={
            "user_id": index + 1,
            "action": "click",
            "page": f"/page/{index % 10}",
            "timestamp": datetime.utcnow().isoformat()
        }
    )


def run_redis_example(host: str, port: int, password: str, table: str, 
                      count: int, batch: bool):
    """Redis 示例"""
    print(f"连接到 Redis {host}:{port}...")
    
    client = RedisStreamClient(host=host, port=port, password=password)
    
    print(f"已连接. 发布 {count} 个事件到表 '{table}'...")
    
    if batch:
        # 批量发送
        records = [generate_record(i) for i in range(count)]
        events = [new_insert_event(table, r, "python-sdk") for r in records]
        
        start = time.time()
        client.publish_batch(events)
        elapsed = time.time() - start
        
        print(f"批量发布 {count} 个事件，耗时 {elapsed:.3f} 秒")
    else:
        # 逐条发送
        for i in range(count):
            record = generate_record(i)
            event = new_insert_event(table, record, "python-sdk")
            msg_id = client.publish(event)
            print(f"发布事件 {i+1}: {event.event_id} -> {msg_id}")
    
    client.close()
    print("完成!")


def run_kafka_example(brokers: str, table: str, count: int, batch: bool):
    """Kafka 示例"""
    broker_list = brokers.split(",")
    print(f"连接到 Kafka {brokers}...")
    
    client = KafkaClient(brokers=broker_list)
    
    print(f"已连接. 发布 {count} 个事件到表 '{table}'...")
    
    if batch:
        # 批量发送
        records = [generate_record(i) for i in range(count)]
        events = [new_insert_event(table, r, "python-sdk") for r in records]
        
        start = time.time()
        client.publish_batch(events)
        elapsed = time.time() - start
        
        print(f"批量发布 {count} 个事件，耗时 {elapsed:.3f} 秒")
    else:
        # 逐条发送
        for i in range(count):
            record = generate_record(i)
            event = new_insert_event(table, record, "python-sdk")
            client.publish(event)
            print(f"发布事件 {i+1}: {event.event_id}")
    
    client.close()
    print("完成!")


def main():
    parser = argparse.ArgumentParser(description="MinIODB Python SDK 示例")
    parser.add_argument("--mode", choices=["redis", "kafka", "grpc"], 
                        default="redis", help="客户端模式")
    parser.add_argument("--redis-host", default="localhost", help="Redis 主机")
    parser.add_argument("--redis-port", type=int, default=6379, help="Redis 端口")
    parser.add_argument("--redis-password", default="", help="Redis 密码")
    parser.add_argument("--kafka-brokers", default="localhost:9092", 
                        help="Kafka brokers (逗号分隔)")
    parser.add_argument("--grpc-addr", default="localhost:8080", help="gRPC 地址")
    parser.add_argument("--table", default="user_events", help="目标表名")
    parser.add_argument("--count", type=int, default=10, help="事件数量")
    parser.add_argument("--batch", action="store_true", help="使用批量模式")
    
    args = parser.parse_args()
    
    if args.mode == "redis":
        run_redis_example(
            args.redis_host, 
            args.redis_port,
            args.redis_password or None,
            args.table, 
            args.count, 
            args.batch
        )
    elif args.mode == "kafka":
        run_kafka_example(args.kafka_brokers, args.table, args.count, args.batch)
    elif args.mode == "grpc":
        print("gRPC 模式需要先生成 protobuf Python 代码")
        print("运行: python -m grpc_tools.protoc -I../../api/proto --python_out=. --grpc_python_out=. miniodb/v1/miniodb.proto")
    else:
        print(f"未知模式: {args.mode}")


if __name__ == "__main__":
    main()

File: examples/sdk/nodejs/miniodb-client.js
#!/usr/bin/env node

/**
 * MinIODB Node.js SDK 示例
 * 
 * 本示例展示如何通过 Redis Streams 或 Kafka 向 MinIODB 发布数据
 * 适用于第三方系统集成，实现业务实时数据同步
 * 
 * 依赖安装:
 *   npm install
 * 
 * 使用方法:
 *   node miniodb-client.js --mode redis    # 使用 Redis Streams
 *   node miniodb-client.js --mode kafka    # 使用 Kafka
 *   node miniodb-client.js --mode grpc     # 使用 gRPC 直连
 */

import { v4 as uuidv4 } from 'uuid';
import { Command } from 'commander';
import Redis from 'ioredis';
import { Kafka } from 'kafkajs';

// ===============================================
// 数据流格式定义（与 MinIODB 保持一致）
// ===============================================

/**
 * 创建数据记录
 * @param {string} id - 记录 ID
 * @param {number} timestamp - 毫秒时间戳
 * @param {Object} payload - 数据负载
 * @returns {Object} DataRecord
 */
function createDataRecord(id, timestamp, payload) {
    return {
        id,
        timestamp,
        payload,
        version: null,
        deleted_at: null
    };
}

/**
 * 创建事件元数据
 * @param {Object} options - 元数据选项
 * @returns {Object} EventMetadata
 */
function createEventMetadata(options = {}) {
    return {
        source: options.source || null,
        batch_id: options.batchId || null,
        partition_key: options.partitionKey || null,
        priority: options.priority || 'normal',
        trace_id: options.traceId || null,
        tags: options.tags || {},
        retry_count: options.retryCount || 0
    };
}

/**
 * 创建数据事件
 * @param {string} eventType - 事件类型: insert, update, delete, batch
 * @param {string} table - 目标表名
 * @param {Array} records - 数据记录列表
 * @param {Object} metadata - 元数据
 * @returns {Object} DataEvent
 */
function createDataEvent(eventType, table, records, metadata = {}) {
    return {
        event_id: `evt_${uuidv4().substring(0, 12)}`,
        event_type: eventType,
        table,
        timestamp: new Date().toISOString(),
        records,
        metadata: createEventMetadata(metadata)
    };
}

/**
 * 创建插入事件
 */
function createInsertEvent(table, record, source = null) {
    return createDataEvent('insert', table, [record], { source });
}

/**
 * 创建批量插入事件
 */
function createBatchEvent(table, records, source = null) {
    return createDataEvent('batch', table, records, { source });
}

// ===============================================
// Redis Streams 客户端
// ===============================================

class RedisStreamClient {
    /**
     * @param {Object} options - Redis 连接选项
     * @param {string} options.host - Redis 主机
     * @param {number} options.port - Redis 端口
     * @param {string} options.password - Redis 密码
     * @param {string} options.streamPrefix - Stream key 前缀
     */
    constructor(options = {}) {
        this.streamPrefix = options.streamPrefix || 'miniodb:stream:';
        
        this.client = new Redis({
            host: options.host || 'localhost',
            port: options.port || 6379,
            password: options.password || undefined,
            retryDelayOnFailover: 100,
            maxRetriesPerRequest: 3
        });
        
        this.client.on('error', (err) => {
            console.error('Redis connection error:', err.message);
        });
    }
    
    /**
     * 发布事件到 Redis Stream
     * @param {Object} event - DataEvent
     * @returns {Promise<string>} Stream Entry ID
     */
    async publish(event) {
        const streamKey = `${this.streamPrefix}${event.table}`;
        
        const fields = [
            'event_id', event.event_id,
            'event_type', event.event_type,
            'table', event.table,
            'timestamp', event.timestamp,
            'data', JSON.stringify(event)
        ];
        
        const entryId = await this.client.xadd(streamKey, '*', ...fields);
        return entryId;
    }
    
    /**
     * 批量发布事件
     * @param {Array} events - DataEvent 列表
     * @returns {Promise<Array>} Entry ID 列表
     */
    async publishBatch(events) {
        const pipeline = this.client.pipeline();
        
        for (const event of events) {
            const streamKey = `${this.streamPrefix}${event.table}`;
            
            pipeline.xadd(streamKey, '*',
                'event_id', event.event_id,
                'event_type', event.event_type,
                'table', event.table,
                'timestamp', event.timestamp,
                'data', JSON.stringify(event)
            );
        }
        
        const results = await pipeline.exec();
        return results.map(([err, id]) => {
            if (err) throw err;
            return id;
        });
    }
    
    /**
     * 关闭连接
     */
    async close() {
        await this.client.quit();
        console.log('Redis connection closed');
    }
}

// ===============================================
// Kafka 客户端
// ===============================================

class KafkaClient {
    /**
     * @param {Object} options - Kafka 连接选项
     * @param {Array<string>} options.brokers - Kafka broker 列表
     * @param {string} options.topicPrefix - Topic 前缀
     * @param {string} options.clientId - 客户端 ID
     */
    constructor(options = {}) {
        this.topicPrefix = options.topicPrefix || 'miniodb-';
        
        this.kafka = new Kafka({
            clientId: options.clientId || 'miniodb-nodejs-sdk',
            brokers: options.brokers || ['localhost:9092'],
            retry: {
                initialRetryTime: 100,
                retries: 3
            }
        });
        
        this.producer = this.kafka.producer({
            allowAutoTopicCreation: true
        });
        
        this.connected = false;
    }
    
    /**
     * 连接到 Kafka
     */
    async connect() {
        if (!this.connected) {
            await this.producer.connect();
            this.connected = true;
            console.log('Connected to Kafka');
        }
    }
    
    /**
     * 发布事件到 Kafka
     * @param {Object} event - DataEvent
     * @returns {Promise<Object>} RecordMetadata
     */
    async publish(event) {
        await this.connect();
        
        const topic = `${this.topicPrefix}${event.table}`;
        
        const result = await this.producer.send({
            topic,
            messages: [{
                key: event.table,
                value: JSON.stringify(event),
                headers: {
                    'event_id': event.event_id,
                    'event_type': event.event_type
                }
            }]
        });
        
        return result;
    }
    
    /**
     * 批量发布事件
     * @param {Array} events - DataEvent 列表
     * @returns {Promise<Array>} RecordMetadata 列表
     */
    async publishBatch(events) {
        await this.connect();
        
        // 按 topic 分组
        const topicMessages = new Map();
        
        for (const event of events) {
            const topic = `${this.topicPrefix}${event.table}`;
            
            if (!topicMessages.has(topic)) {
                topicMessages.set(topic, []);
            }
            
            topicMessages.get(topic).push({
                key: event.table,
                value: JSON.stringify(event),
                headers: {
                    'event_id': event.event_id,
                    'event_type': event.event_type
                }
            });
        }
        
        // 批量发送
        const results = [];
        for (const [topic, messages] of topicMessages) {
            const result = await this.producer.send({ topic, messages });
            results.push(result);
        }
        
        return results;
    }
    
    /**
     * 关闭连接
     */
    async close() {
        if (this.connected) {
            await this.producer.disconnect();
            this.connected = false;
            console.log('Kafka connection closed');
        }
    }
}

// ===============================================
// 工具函数
// ===============================================

/**
 * 生成测试记录
 * @param {number} index - 索引
 * @returns {Object} DataRecord
 */
function generateRecord(index) {
    return createDataRecord(
        `record_${Date.now()}_${index}`,
        Date.now(),
        {
            user_id: index + 1,
            action: 'click',
            page: `/page/${index % 10}`,
            timestamp: new Date().toISOString()
        }
    );
}

/**
 * 生成测试事件
 * @param {string} table - 表名
 * @param {number} index - 索引
 * @returns {Object} DataEvent
 */
function generateEvent(table, index) {
    const record = generateRecord(index);
    return createInsertEvent(table, record, 'nodejs-sdk-example');
}

/**
 * 生成多个测试事件
 * @param {string} table - 表名
 * @param {number} count - 数量
 * @returns {Array} DataEvent 列表
 */
function generateEvents(table, count) {
    const events = [];
    for (let i = 0; i < count; i++) {
        events.push(generateEvent(table, i));
    }
    return events;
}

// ===============================================
// 示例运行函数
// ===============================================

async function runRedisExample(options) {
    console.log(`Connecting to Redis at ${options.redisHost}:${options.redisPort}...`);
    
    const client = new RedisStreamClient({
        host: options.redisHost,
        port: options.redisPort,
        password: options.redisPassword || undefined
    });
    
    try {
        console.log(`Connected. Publishing ${options.count} events to table '${options.table}'...`);
        
        if (options.batch) {
            // 批量发送
            const events = generateEvents(options.table, options.count);
            
            const start = Date.now();
            await client.publishBatch(events);
            const elapsed = Date.now() - start;
            
            console.log(`Published ${options.count} events in batch mode in ${elapsed} ms`);
        } else {
            // 逐条发送
            for (let i = 0; i < options.count; i++) {
                const event = generateEvent(options.table, i);
                const entryId = await client.publish(event);
                console.log(`Published event ${i + 1}: ${event.event_id} -> ${entryId}`);
            }
        }
        
        console.log('Done!');
    } finally {
        await client.close();
    }
}

async function runKafkaExample(options) {
    const brokers = options.kafkaBrokers.split(',');
    console.log(`Connecting to Kafka at ${brokers.join(', ')}...`);
    
    const client = new KafkaClient({ brokers });
    
    try {
        console.log(`Publishing ${options.count} events to table '${options.table}'...`);
        
        if (options.batch) {
            // 批量发送
            const events = generateEvents(options.table, options.count);
            
            const start = Date.now();
            await client.publishBatch(events);
            const elapsed = Date.now() - start;
            
            console.log(`Published ${options.count} events in batch mode in ${elapsed} ms`);
        } else {
            // 逐条发送
            for (let i = 0; i < options.count; i++) {
                const event = generateEvent(options.table, i);
                await client.publish(event);
                console.log(`Published event ${i + 1}: ${event.event_id}`);
            }
        }
        
        console.log('Done!');
    } finally {
        await client.close();
    }
}

async function runGrpcExample(options) {
    console.log('gRPC mode requires proto file compilation.');
    console.log('Please use the Go SDK for gRPC examples, or compile the proto files first:');
    console.log('');
    console.log('  npm install @grpc/proto-loader @grpc/grpc-js');
    console.log('  # Then load the proto file dynamically in your code');
    console.log('');
    process.exit(1);
}

// ===============================================
// 主程序
// ===============================================

const program = new Command();

program
    .name('miniodb-client')
    .description('MinIODB Node.js SDK Example')
    .version('1.0.0')
    .option('-m, --mode <mode>', 'Client mode: redis, kafka, or grpc', 'redis')
    .option('--redis-host <host>', 'Redis host', 'localhost')
    .option('--redis-port <port>', 'Redis port', '6379')
    .option('--redis-password <password>', 'Redis password')
    .option('--kafka-brokers <brokers>', 'Kafka brokers (comma-separated)', 'localhost:9092')
    .option('--grpc-addr <addr>', 'gRPC server address', 'localhost:8080')
    .option('-t, --table <table>', 'Target table name', 'user_events')
    .option('-c, --count <count>', 'Number of events to send', '10')
    .option('-b, --batch', 'Use batch mode', false);

program.parse();

const options = program.opts();
options.count = parseInt(options.count, 10);
options.redisPort = parseInt(options.redisPort, 10);

console.log('MinIODB Node.js SDK Example');
console.log(`Mode: ${options.mode}, Table: ${options.table}, Count: ${options.count}, Batch: ${options.batch}`);
console.log('');

(async () => {
    try {
        switch (options.mode) {
            case 'redis':
                await runRedisExample(options);
                break;
            case 'kafka':
                await runKafkaExample(options);
                break;
            case 'grpc':
                await runGrpcExample(options);
                break;
            default:
                console.error(`Unknown mode: ${options.mode}. Use 'redis', 'kafka', or 'grpc'.`);
                process.exit(1);
        }
    } catch (error) {
        console.error('Error:', error.message);
        process.exit(1);
    }
})();

// ===============================================
// 导出供外部使用
// ===============================================

export {
    createDataRecord,
    createEventMetadata,
    createDataEvent,
    createInsertEvent,
    createBatchEvent,
    RedisStreamClient,
    KafkaClient
};
